### PR TITLE
[multi-asic] Fixed systemd-sonic-generator for multi-asic

### PR DESCRIPTION
**Pull request by anamehra**
Signed-off-by: Anand Mehra <anamehra@cisco.com>

#### Why I did it
1. systemd-sonic-generator limits multi-asic unit file instances to 10 (single digit instance number 0 - 10). This limitation needs to be removed to handle more than 10 asics.
2. MAX_NUM_TARGETS and MAX_NUM_INSTALL_LINES limits to 15 which is not sufficient for systems with more than 15 asics.
3. Inside get_unit_files(), strcmp produce incorrect results due to non null terminated string being compared.

#### How I did it
    1. In function insert_instance_number instance_string was malloced for 2 char
       size which was limiting the instance number value in instance_name to 1 digit.
       Fixed insert_instance_number to use asprintf to generate instancd_name for
       any number of instances. Added _GNU_SOURCE to CFLAGS for asprintf.

    2. Fixed get_unit_files() to use calloc instead of malloc. Uninitialized memory
       was causing incorrect string mismatch error while comparing unit file name
       string.

    3. Increased MAX_NUM_TARGETS and MAX_NUM_INSTALL_LINES values to 48 to handle more
       asic instances.

    4. Added build UT support for systemd-sonic-generator:
        a. Refactor systemd-sonic-generator.c to be used with UT infra.
        b. Added UT infra to run build UT for systemd-sonic-generator
        c. Added functional level and program level UT class and test cases.

#### How to verify it
On a single asic system:
1. Verify Unit file symlinks at following path for single instance system:
/run/systemd/generator/multi-user.target.wants/
/run/systemd/generator/sonic.target.wants/
/run/systemd/generator/swss.service.wants/

On a multi asic system:
1. Verify Unit file symlinks at following path for multi instance system:
/run/systemd/generator/multi-user.target.wants/
/run/systemd/generator/sonic.target.wants/
/run/systemd/generator/swss@n.service.wants/    where n is instance num 0 - (num_asic-1)

Build UT logs:
[systemd-sonic-generator_1.0.0_amd64.deb.log](https://github.com/Azure/sonic-buildimage/files/6497378/systemd-sonic-generator_1.0.0_amd64.deb.log)

Single asic system logs:
[Single_asic.logs.txt](https://github.com/Azure/sonic-buildimage/files/6497388/Single_asic.logs.txt)

Multi asic system logs:
[multi-asic.logs.txt](https://github.com/Azure/sonic-buildimage/files/6504489/multi-asic.logs.txt)

#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012

### CODE REVIEW

File: src/systemd-sonic-generator/ssg-test.cc
Comment (jleveque) [inline src/systemd-sonic-generator/ssg-test.cc:68]: Suggest to add a space before first `:` to align style with other SONiC C++ files.

File: src/systemd-sonic-generator/ssg-test.cc
Comment (jleveque) [inline src/systemd-sonic-generator/ssg-test.cc:102]: Suggest to add a space before first `:` to align style with other SONiC C++ files.

File: src/systemd-sonic-generator/ssg-test.cc
Comment (jleveque) [inline src/systemd-sonic-generator/ssg-test.cc:170]: Suggest to add a space before first `:` to align style with other SONiC C++ files.

File: src/systemd-sonic-generator/systemd-sonic-generator.c
Comment (jleveque) [inline src/systemd-sonic-generator/systemd-sonic-generator.c:22]: Suggest explicitly setting setting to NULL in definition. Can't hurt to be explicit.

File: src/systemd-sonic-generator/systemd-sonic-generator.c
Comment (jleveque) [inline src/systemd-sonic-generator/systemd-sonic-generator.c:27]: Suggest explicitly setting setting to NULL in definition. Can't hurt to be explicit.

File: src/systemd-sonic-generator/systemd-sonic-generator.c
Comment (jleveque) [inline src/systemd-sonic-generator/systemd-sonic-generator.c:32]: Suggest explicitly setting setting to NULL in definition. Can't hurt to be explicit.

File: src/systemd-sonic-generator/systemd-sonic-generator.c
Comment (jleveque) [inline src/systemd-sonic-generator/systemd-sonic-generator.c:37]: Suggest explicitly setting setting to NULL in definition. Can't hurt to be explicit.

**Review by jleveque — CHANGES_REQUESTED** [review]
As comments.

And thank you for adding the unit tests!

**Comment by lguohan** [comment]
@theasianpianist to check

File: src/systemd-sonic-generator/systemd-sonic-generator.c
Comment (theasianpianist) [inline src/systemd-sonic-generator/systemd-sonic-generator.c:213]: Why is this needed?

File: src/systemd-sonic-generator/systemd-sonic-generator.c
Comment (theasianpianist) [inline src/systemd-sonic-generator/systemd-sonic-generator.c:214]: Why are these changes needed?

Comment (anamehra) [reply src/systemd-sonic-generator/systemd-sonic-generator.c:214]: This is to remove extra end of line from the string.
Later below "snprintf(buf, MAX_BUF_SIZE, "%s=%s@%d.%s\n"," adds the needed end of line.
This is to prevent extra empty lines in unit files.

**Review by theasianpianist — DISMISSED** [review]

File: src/systemd-sonic-generator/systemd-sonic-generator.c
Comment (jleveque) [reply src/systemd-sonic-generator/systemd-sonic-generator.c:214]: Couldn't these two lines be simplified to

```
type = strtok_r(NULL, "\n", &save_ptr2);
```

Comment (anamehra) [reply src/systemd-sonic-generator/systemd-sonic-generator.c:214]: The original string may contain multiple service strings separated by white space:
After=unit1.service unit2.service

I was thinking the first statement was doing a split on white space and added Second to remove "\n" if any.
But that whitespace split happens before in while check itself so yes, we don't need two statements. Just one with "\n" is good enough. 
I will make the change, test and upload.

Comment (anamehra) [reply src/systemd-sonic-generator/systemd-sonic-generator.c]: @theasianpianist I have committed the change. You may have to approve again. Thanks!
@jleveque Please approve if you don't have any more comment. Thanks

**Comment by anamehra** [comment]
Hi @lguohan , The build errors reported after new commit does not look like related. Were these observed earlier? May we rerun the builds? Thanks

**Comment by anshuv-mfst** [comment]
Hi @theasianpianist - could you please initiate azp run.

**Comment by anamehra** [comment]
Thanks @theasianpianist for re-executing pipeline.
@anshuv-mfst As all build/test pass, could this be merged or do we need more approvals?

**Comment by anshuv-mfst** [comment]
@theasianpianist - thanks, could you please review and approve, kindly help with merge.

**Review by theasianpianist — APPROVED** [review]

**Comment by lguohan** [comment]
revert this pr, please fix the armhf build.

**Comment by anamehra** [comment]
Hi @lguohan , @theasianpianist , the build issue with armfh was because the static gtest lib was not being pacakged. I have fixed it by using -lgtest which works for both arm64 and armfh builds.

I was wondering if I should split this PR in two PRs, one for actual systemd-sonic-generator changes to required multi-asic enhancements and the other for UT code enhancement only. Please advise.

**Comment by theasianpianist** [comment]
@anamehra unless @lguohan has a preference, I am fine with either

**Comment by anamehra** [comment]
@theasianpianist Created #7954 with only change in Makefile for -lgtest

**Pull request merged**
